pkg/kubelet/apis/pods/server.go

```diff
@@ -0,0 +1,508 @@
+
+	clientChannel := make(chan PodWatchEvent, 100)
+	s.broadcaster.Register(clientChannel)
+	defer func() {
+		s.broadcaster.Unregister(clientChannel)
+		logger.Info("Watch client disconnected", "client", clientAddr)
+	}()
+
+	fieldMask, err := getFieldMaskFromContext(stream.Context())
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "invalid field mask: %v", err)
+	}
+
+	// Validate the field mask before starting the watch.
+	if _, err := applyFieldMaskToPod(&v1.Pod{}, fieldMask); err != nil {
+		return status.Errorf(codes.InvalidArgument, "invalid field mask: %v", err)
+	}
+
+	// Send initial ADDED events
+	initialPods := s.List()
+	for _, p := range initialPods {
+		maskedPod, err := applyFieldMaskToPod(p, fieldMask)
+		if err != nil {
+			// Should not happen, the field mask was already validated.
+			logger.Error(err, "Error applying field mask to initial watch event pod")
+			metrics.PodWatchEventsDroppedTotal.Inc()
+			continue
+		}
+		podBytes, err := maskedPod.Marshal()
+		if err != nil {
+			logger.Error(err, "Error marshalling initial watch event pod")
+			metrics.PodWatchEventsDroppedTotal.Inc()
+			continue
+		}
+		if err := stream.Send(&podsv1alpha1.WatchPodsEvent{
+			Type: podsv1alpha1.EventType_ADDED,
+			Pod:  podBytes,
+		}); err != nil {
+			logger.Error(err, "Error sending initial watch event")
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			logger.Info("Watch context cancelled", "client", clientAddr)
+			return stream.Context().Err()
+		case event := <-clientChannel:
+			maskedPod, err := applyFieldMaskToPod(event.Pod, fieldMask)
+			if err != nil {
+				logger.Error(err, "Error applying field mask to watch event pod")
+				metrics.PodWatchEventsDroppedTotal.Inc()
+				continue
+			}
+			podBytes, err := maskedPod.Marshal()
+			if err != nil {
+				logger.Error(err, "Error marshalling watch event pod")
+				metrics.PodWatchEventsDroppedTotal.Inc()
+				continue
+			}
+			if err := stream.Send(&podsv1alpha1.WatchPodsEvent{
+				Type: convertWatchEventType(event.Type),
+				Pod:  podBytes,
+			}); err != nil {
+				logger.Error(err, "Error sending watch event to client", "client", clientAddr)
+				return err
+			}
+		}
+	}
+}
+
+func convertWatchEventType(watchType watch.EventType) podsv1alpha1.EventType {
+	switch watchType {
+	case watch.Added:
+		return podsv1alpha1.EventType_ADDED
+	case watch.Modified:
+		return podsv1alpha1.EventType_MODIFIED
+	case watch.Deleted:
+		return podsv1alpha1.EventType_DELETED
+	default:
+		return podsv1alpha1.EventType_ADDED
+	}
+}
+
+// getFieldMaskFromContext extracts the field mask from the gRPC metadata.
+func getFieldMaskFromContext(ctx context.Context) (*fieldmaskpb.FieldMask, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, nil
+	}
+	fieldMasks := md.Get(FieldMaskMetadataKey)
+	if len(fieldMasks) == 0 {
+		return nil, nil
+	}
+	if len(fieldMasks) > 1 {
```

Comment:
why? It is comma separated list, why not allow to split it into multiple headers. At least in Http it is a common practice and semantically interchangeable - list in one header comma separated vs. multiple headers. Furthermore, for http some client libraries are opinionated on what to do with multi-value comma separated headers